### PR TITLE
Replace concrete usage with protocol: `ProductFormDataModel`, `ProductFormActionsFactoryProtocol`, and `TaxClassRequestable`

### PR DESCRIPTION
**Pull request by jaclync**
Prerequisite for #2085 

This PR abstracts parts of concrete usage with protocol in the product form and its related classes, so that `ProductVariation` can be supported for these classes in a future subtask. These changes should be pretty straightforward. I left the more complicated protocol that abstracts the current `ProductFormViewModel` for a future subtask. In addition, I left the `Product` protocol replacement in price and inventory settings for later since these 2 screens have UI changes between `Product` and `ProductVariation` (p91TBi-33Q-p2).

## Changes

- Created `ProductFormActionsFactoryProtocol` protocol that the existing `ProductFormActionsFactory` conforms to, and updated its existing usage
- Created `ProductFormDataModel` protocol that abstracts the `Product` usage in product form. It contains all the necessary fields for `ProductFormViewController` and its navigated view controllers. Replaced `Product` with `ProductFormDataModel` in product form related classes where replacement is straightforward
- Created `TaxClassRequestable` protocol that abstracts the `Product` usage in `TaxClassAction.requestMissingTaxClasses` within price settings. Updated existing usage in price settings classes
- Removed unused `product: Product` associated value in `ProductFormSection.PrimaryFieldRow.images` enum case
- Removed redundant extension helper `isShippingEnabled` in Yosemite layer and moved its unit tests from Yosemite to app layer.

## Testing

No changes in app behavior are expected, feel free to test for regression with the following steps:

- Go to the Products tab
- Tap on a simple product
- Tap on the description row, and make sure the editing behavior stays the same
- Go back to the product details
- Tap on the price settings row, and make sure the editing behavior stays the same especially on the tax class selector
- Go back to the product details
- Tap on the shipping settings row, and make sure the editing behavior stays the same especially on the shipping class selector
- Go back to the product details
- Tap on the "+" cell in the images header, and make sure image editing behavior stays the same

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        fullDescription
+    }
+
+    var shortDescription: String? {
+        briefDescription
+    }
+
+    var stockStatus: ProductStockStatus {
+        productStockStatus
+    }
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift:80]: Nit: It looks a bit odd that these don't use the return keyword but the ones above do. Do you think we should make it consistent?

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    /// Returns `ProductTaxStatus` given the raw value from `taxStatusKey` field.
+    var productTaxStatus: ProductTaxStatus {
+        return ProductTaxStatus(rawValue: taxStatusKey)
+    }
+
+    /// Returns `ProductBackordersSetting` given the raw value from `backordersKey` field.
+    var backordersSetting: ProductBackordersSetting {
+        return ProductBackordersSetting(rawValue: backordersKey)
+    }
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift:66]: Q: is it worth to add unit tests to these functions?

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift:80]: yeaa I'm all for consistency! update in https://github.com/woocommerce/woocommerce-ios/pull/2552/commits/3f200d10f2fe67ef154aae7ad9807cf636d2031b to remove unnecessary `return` in the ones above (one line return)

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormDataModel.swift]: yes! we have some existing unit tests and I just added some:

- `trimmedFullDescription`: [`Product_ProductFormTests.testTrimmedFullDescriptionWithLeadingNewLinesAndHTMLTags`](https://github.com/woocommerce/woocommerce-ios/blob/09573496fe95d51dbeb4125a8fb57ea2e76cfd2c/WooCommerce/WooCommerceTests/ViewRelated/Products/Product+ProductFormTests.swift#L10-L15)
- `isShippingEnabled`: [`Product_ShippingTests`](https://github.com/woocommerce/woocommerce-ios/blob/fcc5b976b691127efbbd215a5cef99577ee28242/WooCommerce/WooCommerceTests/Extensions/Product%2BShippingTests.swift)
- `productTaxStatus` and `backordersSetting`: just added in https://github.com/woocommerce/woocommerce-ios/pull/2552/commits/27ac6231e47020fa94dd7d95f22750cd757582bd

**Comment by jaclync** [comment]
Thanks for the suggestions Ernesto! updated the PR based on your feedback 😄 merging this now!

**Pull request merged**